// @livestore/adapter-cloudflare@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/make-client-durable-object.js
import { createStore, provideOtel } from '@livestore/livestore';
import { makeDoRpcSync } from '@livestore/sync-cf/client';
import { Effect, Logger, LogLevel, Scope } from '@livestore/utils/effect';
import { makeAdapter } from "./make-adapter.js";
export const createStoreDo = ({ schema, storeId, clientId, sessionId, storage, syncBackendDurableObject, durableObjectId, bindingName, livePull = false, }) => Effect.gen(function* () {
    const scope = yield* Scope.make();
    const adapter = makeAdapter({
        clientId,
        sessionId,
        storage,
        syncOptions: {
            backend: makeDoRpcSync({
                syncBackendStub: syncBackendDurableObject,
                durableObjectContext: { bindingName, durableObjectId },
            }),
            livePull, // Uses DO RPC callbacks for reactive pull
            // backend: makeHttpSync({ url: `http://localhost:8787`, livePull: { pollInterval: 500 } }),
            initialSyncOptions: { _tag: 'Blocking', timeout: 500 },
            // backend: makeWsSyncProviderClient({ durableObject: syncBackendDurableObject }),
        },
    });
    return yield* createStore({ schema, adapter, storeId }).pipe(Scope.extend(scope), provideOtel({}));
});
export const createStoreDoPromise = (options) => createStoreDo(options).pipe(Logger.withMinimumLogLevel(LogLevel.Debug), Effect.provide(Logger.consoleWithThread('DoClient')), Effect.tapCauseLogPretty, Effect.runPromise);
