// @livestore/adapter-cloudflare@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/polyfill.js
/// <reference lib="dom" />
// TODO remove all unused polyfills once we're closer to the release
globalThis.performance = globalThis.performance ?? {};
globalThis.performance.mark = globalThis.performance.mark ?? (() => { });
globalThis.performance.measure = globalThis.performance.measure ?? (() => { });
globalThis.performance.now = globalThis.performance.now ?? (() => -1);
if (typeof globalThis.location === 'undefined') {
    globalThis.location = {
        href: 'https://worker.cloudflare.com/',
        origin: 'https://worker.cloudflare.com',
        protocol: 'https:',
        host: 'worker.cloudflare.com',
        hostname: 'worker.cloudflare.com',
        port: '',
        pathname: '/',
        search: '',
        hash: '',
    };
}
if (typeof globalThis.document === 'undefined') {
    globalThis.document = {
        createElement: () => ({ href: '', pathname: '', search: '', origin: '' }),
        head: { appendChild: () => { } },
    };
}
// WeakRef polyfill for Cloudflare Workers
if (typeof WeakRef === 'undefined') {
    // @ts-expect-error
    globalThis.WeakRef = class WeakRef {
        target;
        constructor(target) {
            this.target = target;
        }
        deref() {
            return this.target;
        }
    };
}
export {};
