// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/common/sync-message-types.js
import { SyncBackend, UnexpectedError } from '@livestore/common';
import { EventSequenceNumber, LiveStoreEvent } from '@livestore/common/schema';
import { Schema } from '@livestore/utils/effect';
/**
 * Common sync message types shared between different transport modes (WS, HTTP, RPC)
 *
 * These are the canonical message definitions used across all transport implementations.
 */
export const SyncMetadata = Schema.TaggedStruct('SyncMessage.SyncMetadata', {
    /** ISO date format */
    createdAt: Schema.String,
}).annotations({ title: '@livestore/sync-cf:SyncMetadata' });
export const PullRequest = Schema.Struct({
    /** Omitting the cursor will start from the beginning */
    cursor: Schema.optional(EventSequenceNumber.GlobalEventSequenceNumber),
}).annotations({ title: '@livestore/sync-cf:PullRequest' });
export const PullResponse = Schema.Struct({
    batch: Schema.Array(Schema.Struct({
        eventEncoded: LiveStoreEvent.AnyEncodedGlobal,
        metadata: Schema.Option(SyncMetadata),
    })),
    pageInfo: SyncBackend.PullResPageInfo,
}).annotations({ title: '@livestore/sync-cf:PullResponse' });
export const PushRequest = Schema.Struct({
    batch: Schema.Array(LiveStoreEvent.AnyEncodedGlobal),
}).annotations({ title: '@livestore/sync-cf:PushRequest' });
export const PushAck = Schema.Struct({}).annotations({
    title: '@livestore/sync-cf:PushAck',
});
export const InvalidParentEventNumber = Schema.TaggedStruct('SyncMessage.SyncError.InvalidParentEventNumber', {
    expected: EventSequenceNumber.GlobalEventSequenceNumber,
    received: EventSequenceNumber.GlobalEventSequenceNumber,
}).annotations({ title: '@livestore/sync-cf:InvalidParentEventNumber' });
export class SyncError extends Schema.TaggedError()('SyncMessage.SyncError', {
    cause: Schema.Union(UnexpectedError, InvalidParentEventNumber),
    storeId: Schema.optional(Schema.String),
}, { title: '@livestore/sync-cf:SyncError' }) {
}
export const Ping = Schema.TaggedStruct('SyncMessage.Ping', {}).annotations({ title: '@livestore/sync-cf:Ping' });
export const Pong = Schema.TaggedStruct('SyncMessage.Pong', {}).annotations({ title: '@livestore/sync-cf:Pong' });
// Admin operations
export const AdminResetRoomRequest = Schema.TaggedStruct('SyncMessage.AdminResetRoomRequest', {
    adminSecret: Schema.String,
}).annotations({ title: '@livestore/sync-cf:AdminResetRoomRequest' });
export const AdminResetRoomResponse = Schema.TaggedStruct('SyncMessage.AdminResetRoomResponse', {}).annotations({
    title: '@livestore/sync-cf:AdminResetRoomResponse',
});
export const AdminInfoRequest = Schema.TaggedStruct('SyncMessage.AdminInfoRequest', {
    adminSecret: Schema.String,
}).annotations({ title: '@livestore/sync-cf:AdminInfoRequest' });
export const AdminInfoResponse = Schema.TaggedStruct('SyncMessage.AdminInfoResponse', {
    info: Schema.Struct({
        durableObjectId: Schema.String,
    }),
}).annotations({ title: '@livestore/sync-cf:AdminInfoResponse' });
export const BackendToClientMessage = Schema.Union(PullResponse, PushAck, SyncError, Pong, AdminResetRoomResponse, AdminInfoResponse);
export const ClientToBackendMessage = Schema.Union(PullRequest, PushRequest, Ping, AdminResetRoomRequest, AdminInfoRequest);
export const Message = Schema.Union(BackendToClientMessage, ClientToBackendMessage);
