// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/sync-impl/transport/http.js
import { InvalidPullError, InvalidPushError, SyncBackend, UnexpectedError } from '@livestore/common';
import { shouldNeverHappen } from '@livestore/utils';
import { Effect, HttpClient, HttpClientRequest, Layer, Option, RpcClient, RpcSerialization, Schedule, SubscriptionRef, } from '@livestore/utils/effect';
import { nanoid } from '@livestore/utils/nanoid';
import { SyncHttpRpc } from "../../common/http-rpc-schema.js";
export const makeHttpSync = (options) => ({ storeId, payload }) => Effect.gen(function* () {
    // Based on ping responses
    const isConnected = yield* SubscriptionRef.make(false);
    // Setup HTTP RPC Protocol
    const HttpProtocolLive = RpcClient.layerProtocolHttp({
        url: `${options.baseUrl}/http-rpc`,
        transformClient: HttpClient.mapRequest((request) => request.pipe(HttpClientRequest.setHeaders({
            ...options.headers,
            // Used in CF Worker to identify the store (additionally to storeId embedded in the RPC requests)
            'x-livestore-store-id': storeId,
        }))),
    }).pipe(Layer.provide(RpcSerialization.layerJson));
    const rpcClient = yield* RpcClient.make(SyncHttpRpc).pipe(Effect.provide(HttpProtocolLive));
    const pingTimeout = options.ping?.requestTimeout ?? 10_000;
    const ping = Effect.gen(function* () {
        const requestId = nanoid();
        const res = yield* rpcClient.SyncHttpRpc.Ping({ storeId, payload, requestId });
        if (res.requestId !== requestId) {
            return shouldNeverHappen(`Ping response requestId ${res.requestId} does not match expected requestId ${requestId}`);
        }
        yield* SubscriptionRef.set(isConnected, true);
    }).pipe(UnexpectedError.mapToUnexpectedError, Effect.timeout(pingTimeout), Effect.catchTag('TimeoutException', () => SubscriptionRef.set(isConnected, false)));
    const pingInterval = options.ping?.requestInterval ?? 10_000;
    // Automatically ping the server to keep the connection alive
    yield* ping.pipe(Effect.repeat(Schedule.spaced(pingInterval)), Effect.tapCauseLogPretty, Effect.forkScoped);
    // Helps already establish a TCP connection to the server
    const connect = ping.pipe(UnexpectedError.mapToUnexpectedError);
    const pull = (args) => Effect.gen(function* () {
        const cursor = Option.getOrUndefined(args)?.cursor.global;
        const requestId = nanoid();
        return yield* rpcClient.SyncHttpRpc.Pull({
            storeId,
            payload,
            requestId,
            cursor,
        }).pipe(Effect.tapCauseLogPretty, Effect.mapError((cause) => new InvalidPullError({ cause })), Effect.map((pullRes) => ({
            batch: pullRes.batch,
            remaining: pullRes.remaining,
        })), Effect.withSpan('http-sync-client:pull'));
    });
    const pushSemaphore = yield* Effect.makeSemaphore(1);
    const push = (batch) => Effect.gen(function* () {
        if (batch.length === 0) {
            return;
        }
        const requestId = nanoid();
        yield* rpcClient.SyncHttpRpc.Push({
            storeId,
            payload,
            requestId,
            batch,
        });
    }).pipe(pushSemaphore.withPermits(1), Effect.mapError((error) => new InvalidPushError({ reason: { _tag: 'Unexpected', cause: error } })), Effect.withSpan('http-sync-client:push'));
    return SyncBackend.of({
        connect,
        isConnected,
        pull,
        push,
        ping,
        metadata: {
            name: '@livestore/cf-sync-http',
            description: 'LiveStore sync backend implementation using HTTP RPC',
            protocol: 'http',
            url: options.baseUrl,
        },
    });
});
