// @livestore/common@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/sync/errors.js
import { Schema } from '@livestore/utils/effect';
import { EventSequenceNumber } from "../schema/mod.js";
export class IsOfflineError extends Schema.TaggedError()('IsOfflineError', {
    cause: Schema.Defect,
}) {
}
// TODO gt rid of this error in favour of SyncError
export class InvalidPushError extends Schema.TaggedError()('InvalidPushError', {
    reason: Schema.Union(Schema.TaggedStruct('Unexpected', {
        cause: Schema.Defect,
    }), Schema.TaggedStruct('ServerAhead', {
        minimumExpectedNum: Schema.Number,
        providedNum: Schema.Number,
    })),
}) {
}
// TODO gt rid of this error in favour of SyncError
export class InvalidPullError extends Schema.TaggedError()('InvalidPullError', {
    cause: Schema.Defect,
}) {
}
// TODO gt rid of this error in favour of SyncError
export class LeaderAheadError extends Schema.TaggedError()('LeaderAheadError', {
    minimumExpectedNum: EventSequenceNumber.EventSequenceNumber,
    providedNum: EventSequenceNumber.EventSequenceNumber,
    /** Generation number the client session should use for subsequent pushes */
    // nextGeneration: Schema.Number,
}) {
}
export const of = (obj) => {
    return obj;
};
