// @livestore/common@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/schema/state/sqlite/column-annotations.js
import { dual } from '@livestore/utils/effect';
export const PrimaryKeyId = Symbol.for('livestore/state/sqlite/annotations/primary-key');
export const ColumnType = Symbol.for('livestore/state/sqlite/annotations/column-type');
export const Default = Symbol.for('livestore/state/sqlite/annotations/default');
export const AutoIncrement = Symbol.for('livestore/state/sqlite/annotations/auto-increment');
export const Unique = Symbol.for('livestore/state/sqlite/annotations/unique');
// export const Check = Symbol.for('livestore/state/sqlite/annotations/check')
/*
Here are the knobs you can turn per-column when you CREATE TABLE (or ALTER TABLE … ADD COLUMN) in SQLite:
•	Declared type / affinity – INTEGER, TEXT, REAL, BLOB, NUMERIC, etc.  ￼
•	NULL vs NOT NULL – disallow NULL on inserts/updates.  ￼
•	PRIMARY KEY – makes the column the rowid (and, if the type is INTEGER, it enables rowid-based auto- numbering). Add the optional AUTOINCREMENT keyword if you need monotonic, never-reused ids.  ￼
•	UNIQUE – enforces per-column uniqueness.  ￼
•	DEFAULT <expr> – literal, function (e.g. CURRENT_TIMESTAMP), or parenthesised expression; since 3.46 you can even default to large hex blobs.  ￼
•	CHECK (<expr>) – arbitrary boolean expression evaluated on write.  ￼
•	COLLATE <name> – per-column collation sequence for text comparison.  ￼
•	REFERENCES tbl(col) [ON UPDATE/DELETE …] – column-local foreign key with its own cascade / restrict / set-null rules.  ￼
•	GENERATED ALWAYS AS (<expr>) [VIRTUAL | STORED] – computed columns (since 3.31).  ￼
•	CONSTRAINT name … – optional label in front of any of the above so you can refer to it in error messages or when dropping/recreating schemas.
*/
/**
 * Adds a primary key annotation to a schema.
 */
export const withPrimaryKey = (schema) => schema.annotations({ [PrimaryKeyId]: true });
/**
 * Adds a column type annotation to a schema.
 */
export const withColumnType = dual(2, (schema, type) => {
    validateSchemaColumnTypeCompatibility(schema, type);
    return schema.annotations({ [ColumnType]: type });
});
/**
 * Adds an auto-increment annotation to a schema.
 */
export const withAutoIncrement = (schema) => schema.annotations({ [AutoIncrement]: true });
/**
 * Adds a unique constraint annotation to a schema.
 */
export const withUnique = (schema) => schema.annotations({ [Unique]: true });
/**
 * Adds a default value annotation to a schema.
 */
export const withDefault = dual(2, (schema, value) => schema.annotations({ [Default]: value }));
/**
 * Validates that a schema is compatible with the specified SQLite column type
 */
const validateSchemaColumnTypeCompatibility = (_schema, _columnType) => {
    // TODO actually implement this
};
