// @livestore/common@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/errors.js
import { Effect, Schema, Stream } from '@livestore/utils/effect';
export class UnexpectedError extends Schema.TaggedError()('LiveStore.UnexpectedError', {
    cause: Schema.Defect,
    note: Schema.optional(Schema.String),
    payload: Schema.optional(Schema.Any),
}) {
    static mapToUnexpectedError = (effect) => effect.pipe(Effect.mapError((cause) => (Schema.is(UnexpectedError)(cause) ? cause : new UnexpectedError({ cause }))), Effect.catchAllDefect((cause) => new UnexpectedError({ cause })));
    static mapToUnexpectedErrorStream = (stream) => stream.pipe(Stream.mapError((cause) => (Schema.is(UnexpectedError)(cause) ? cause : new UnexpectedError({ cause }))));
}
export class SyncError extends Schema.TaggedError()('LiveStore.SyncError', {
    cause: Schema.Defect,
}) {
}
export class MaterializerHashMismatchError extends Schema.TaggedError()('LiveStore.MaterializerHashMismatchError', {
    eventName: Schema.String,
    note: Schema.optionalWith(Schema.String, {
        default: () => 'Please make sure your event materializer is a pure function without side effects.',
    }),
}) {
}
export class IntentionalShutdownCause extends Schema.TaggedError()('LiveStore.IntentionalShutdownCause', {
    reason: Schema.Literal('devtools-reset', 'devtools-import', 'adapter-reset', 'manual'),
}) {
}
export class StoreInterrupted extends Schema.TaggedError()('LiveStore.StoreInterrupted', {
    reason: Schema.String,
}) {
}
export class SqliteError extends Schema.TaggedError()('LiveStore.SqliteError', {
    query: Schema.optional(Schema.Struct({
        sql: Schema.String,
        bindValues: Schema.Union(Schema.Record({ key: Schema.String, value: Schema.Any }), Schema.Array(Schema.Any)),
    })),
    /** The SQLite result code */
    // code: Schema.optional(Schema.Number),
    // Added string support for Expo SQLite (we should refactor this to have a unified error type)
    code: Schema.optional(Schema.Union(Schema.Number, Schema.String)),
    /** The original SQLite3 error */
    cause: Schema.Defect,
    note: Schema.optional(Schema.String),
}) {
}
