// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/sync-impl/transport-factory.js
import { SyncBackend } from '@livestore/common';
import { Effect } from '@livestore/utils/effect';
import { makeHttpSync } from "./transport/http.js";
import { makeDoRpcSync } from "./transport/do-rpc.js";
import { makeCfSync } from "./ws-impl.js";
/**
 * Factory function to create sync backends with different transport modes
 */
export const makeSyncBackend = (config) => {
    switch (config.mode) {
        case 'http': {
            const httpOptions = {
                url: config.url,
                headers: config.headers,
            };
            return makeHttpSync(httpOptions);
        }
        case 'ws': {
            const wsOptions = {
                url: config.url,
                webSocketFactory: config.webSocketFactory,
            };
            return makeCfSync(wsOptions);
        }
        case 'rpc': {
            const rpcOptions = {
                durableObjectStub: config.durableObjectStub,
                clientId: config.clientId,
                durableObjectId: config.durableObjectId,
            };
            return makeDoRpcSync(rpcOptions);
        }
        default: {
            const exhaustiveCheck = config;
            throw new Error(`Unsupported transport mode: ${exhaustiveCheck.mode}`);
        }
    }
};
/**
 * Utility function to determine transport mode from URL
 */
export const inferTransportMode = (url) => {
    try {
        const parsedUrl = new URL(url);
        if (parsedUrl.protocol === 'ws:' || parsedUrl.protocol === 'wss:') {
            return 'ws';
        }
        if (parsedUrl.protocol === 'http:' || parsedUrl.protocol === 'https:') {
            // Check for WebSocket upgrade endpoints
            if (parsedUrl.pathname.includes('websocket') || parsedUrl.pathname.includes('ws')) {
                return 'ws';
            }
            return 'http';
        }
        // Default to HTTP for other protocols
        return 'http';
    }
    catch {
        // If URL parsing fails, default to HTTP
        return 'http';
    }
};
/**
 * Convenience function to create sync backend from URL with automatic transport detection
 */
export const makeSyncBackendFromUrl = (url, options = {}) => {
    const mode = inferTransportMode(url);
    const config = {
        url,
        mode,
        ...options,
    };
    return makeSyncBackend(config);
};
