// @livestore/livestore@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/utils/dev.js
import { prettyBytes } from '@livestore/utils';
import { Effect } from '@livestore/utils/effect';
export const downloadBlob = (data, fileName, mimeType = 'application/octet-stream') => {
    const blob = data instanceof Blob ? data : new Blob([data], { type: mimeType });
    const url = window.URL.createObjectURL(blob);
    downloadURL(url, fileName);
    setTimeout(() => window.URL.revokeObjectURL(url), 1000);
};
export const downloadURL = (data, fileName) => {
    const a = document.createElement('a');
    a.href = data;
    a.download = fileName;
    document.body.append(a);
    a.style.display = 'none';
    a.click();
    a.remove();
};
export const exposeDebugUtils = () => {
    globalThis.__debugLiveStoreUtils = {
        downloadBlob,
        runSync: (effect) => Effect.runSync(effect),
        runFork: (effect) => Effect.runFork(effect),
        dumpDb: (db) => {
            const tables = db.select(`SELECT name FROM sqlite_master WHERE type='table'`);
            for (const table of tables) {
                const rows = db.select(`SELECT * FROM ${table.name}`);
                console.log(`Table: ${table.name} (${prettyBytes(table.name.length)}, ${rows.length} rows)`);
                console.table(rows);
            }
        },
    };
};
