// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/cf-worker/shared.js
import { Effect, Schema, UrlParams } from '@livestore/utils/effect';
import { SearchParamsSchema, SyncMessage } from "../common/mod.js";
/**
 * Needs to be bumped when the storage format changes (e.g. eventlogTable schema changes)
 *
 * Changing this version number will lead to a "soft reset".
 */
export const PERSISTENCE_FORMAT_VERSION = 7;
export const DEFAULT_SYNC_DURABLE_OBJECT_NAME = 'SYNC_BACKEND_DO';
export const encodeOutgoingMessage = Schema.encodeSync(Schema.parseJson(SyncMessage.BackendToClientMessage));
export const encodeIncomingMessage = Schema.encodeSync(Schema.parseJson(SyncMessage.ClientToBackendMessage));
export const getSyncRequestSearchParams = (request) => {
    const url = new URL(request.url);
    const urlParams = UrlParams.fromInput(url.searchParams);
    const paramsResult = UrlParams.schemaStruct(SearchParamsSchema)(urlParams).pipe(Effect.option, Effect.runSync);
    return paramsResult;
};
export const PULL_CHUNK_SIZE = 100;
export const WebSocketAttachmentSchema = Schema.parseJson(Schema.Struct({
    // Same across all websocket connections
    storeId: Schema.String,
    // Different for each websocket connection
    payload: Schema.optional(Schema.JsonValue),
    pullRequestIds: Schema.Array(Schema.String),
}));
