// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/cf-worker/do/transport/do-rpc.js
import { toDurableObjectHandler } from '@livestore/common-cf';
import { Effect, Headers, HttpServer, Layer, Logger, LogLevel, Option, RpcSerialization, Schedule, Stream, } from '@livestore/utils/effect';
import { SyncDoRpc } from "../../../common/do-rpc-schema.js";
import { SyncMessage } from "../../../common/mod.js";
import { makeEndingPullStream } from "../pull.js";
import { makePush } from "../push.js";
import { makeStorage } from "../sync-storage.js";
export const createDoRpcHandler = (options) => Effect.gen(this, function* () {
    const { ctx, env, rpcSubscriptions, currentHeadRef, payload, ensureStorageCache } = options;
    const RpcLive = SyncDoRpc.toLayer({
        'SyncDoRpc.Ping': (_req) => {
            return Effect.succeed(SyncMessage.Pong.make({}));
        },
        'SyncDoRpc.Pull': (req) => Effect.gen(this, function* () {
            yield* ensureStorageCache(req.storeId);
            const pull = makeEndingPullStream({
                storage: makeStorage(ctx, env, req.storeId),
                doOptions: options.doOptions,
                storeId: req.storeId,
                payload: req.payload,
                emitEmptyBatch: false,
            });
            return pull(req);
        }).pipe(Stream.unwrap, Stream.mapError((cause) => SyncMessage.SyncError.make({ cause, storeId: req.storeId }))),
        'SyncDoRpc.Push': (req) => Effect.gen(this, function* () {
            yield* ensureStorageCache(req.storeId);
            const push = makePush({
                storage: makeStorage(ctx, env, req.storeId),
                ctx: ctx,
                env,
                currentHeadRef,
                storeId: req.storeId,
                payload: undefined,
                rpcSubscriptions: rpcSubscriptions,
                options: options.doOptions,
            });
            return yield* push(req);
        }),
        'SyncDoRpc.Subscribe': (req, headers) => Effect.gen(this, function* () {
            yield* ensureStorageCache(req.storeId);
            // const subscribe = makeSubscribe({ storage: makeStorage(ctx, env, req.storeId), requestId: 'ping' })
            rpcSubscriptions.set(req.callerContext.durableObjectId, {
                clientId: req.clientId,
                storeId: req.storeId,
                payload: req.payload,
                subscribedAt: Date.now(),
                requestId: Headers.get(headers, 'x-rpc-request-id').pipe(Option.getOrThrow),
                callerContext: req.callerContext,
            });
            // const res = yield* subscribe(req)
            // TODO get rid of "hard coded" 1s interval in favour of proper reactive subscription
            return Stream.succeed('ok').pipe(Stream.repeat(Schedule.spaced(1000)));
        }).pipe(Stream.unwrap),
        'SyncDoRpc.Unsubscribe': (req) => Effect.gen(this, function* () {
            yield* ensureStorageCache(req.storeId);
            rpcSubscriptions.delete(req.durableObjectId);
        }),
    });
    const handler = toDurableObjectHandler(SyncDoRpc, {
        layer: Layer.mergeAll(RpcLive, RpcSerialization.layerJson, HttpServer.layerContext).pipe(Layer.provide(Logger.prettyWithThread('SyncDo')), Layer.provide(Logger.minimumLogLevel(LogLevel.Debug))),
    });
    return yield* handler(payload);
}).pipe(Effect.withSpan('createDoRpcHandler'));
