// @livestore/sync-cf@0.0.0-snapshot-8452e32b7fbfc129741b253b9c853f866b52129f/dist/cf-worker/do/transport/http-rpc-server.js
import { Effect, HttpApp, Layer, RpcSerialization, RpcServer, Stream } from '@livestore/utils/effect';
import { SyncHttpRpc } from "../../../common/http-rpc-schema.js";
import * as SyncMessage from "../../../common/sync-message-types.js";
import { makeEndingPullStream } from "../pull.js";
import { makePush } from "../push.js";
/**
 * Creates an HTTP RPC web handler for the sync provider
 */
export const createHttpRpcHandler = (options) => Effect.gen(function* () {
    const handlerLayer = createHttpRpcLayer(options);
    const httpApp = RpcServer.toHttpApp(SyncHttpRpc).pipe(Effect.provide(handlerLayer));
    const webHandler = yield* httpApp.pipe(Effect.map(HttpApp.toWebHandler));
    return yield* Effect.promise(() => webHandler(options.request)).pipe(Effect.timeout(10000));
}).pipe(Effect.withSpan('createHttpRpcHandler'));
const createHttpRpcLayer = (options) => 
// TODO implement admin requests
SyncHttpRpc.toLayer({
    'SyncHttpRpc.Pull': (req) => Effect.gen(function* () {
        const storage = yield* options.makeStorage(req.storeId);
        const pull = makeEndingPullStream({
            storage,
            doOptions: options.doOptions,
            storeId: req.storeId,
            payload: req.payload,
        });
        return pull(req);
    }).pipe(Stream.unwrap, Stream.mapError((cause) => SyncMessage.SyncError.make({ cause, storeId: req.storeId }))),
    'SyncHttpRpc.Push': (req) => Effect.gen(function* () {
        const storage = yield* options.makeStorage(req.storeId);
        const push = makePush({
            storage,
            ctx: options.ctx,
            env: options.env,
            currentHeadRef: options.currentHeadRef,
            storeId: req.storeId,
            payload: undefined,
            rpcSubscriptions: options.rpcSubscriptions,
            options: options.doOptions,
        });
        return yield* push(req);
    }),
    'SyncHttpRpc.Ping': () => Effect.succeed(SyncMessage.Pong.make({})),
}).pipe(Layer.provideMerge(RpcServer.layerProtocolHttp({ path: '/http-rpc' })), Layer.provideMerge(RpcSerialization.layerJson));
